### ISSUE TITLE
IIM roadmap

### ISSUE THREAD

**Issue by drwells** [drwells/fiddle#159]
From a discussion today. This is not something I have time to complete myself but I still want to write down some notes.

The fundamental geometric problem we need to solve is ray intersection: we need to find 1) the points where a finite difference stencil intersects the mesh and 2) the corresponding unique element. If we get those then we can start building IIM algorithms.

There's a lot of interesting things to investigate:
1. how do we handle small elements which do not have any corresponding finite difference stencils?
2. how can we do higher-order elements where the intersection code is really hard?
3. what happens if a stencil intersects multiple elements?

@ryanleaf1000 A good place to start would be to look at how to intersect rays and elements in serial: i.e., given a `fdl::PatchMap`, compute a new structure describing where the intersections are with SAMRAI indices, deal.II elements, points on the structure, and points in the reference configuration. We should get this working before doing any physics. This is similar to https://github.com/drwells/fiddle/blob/057b76976b93b3f5c140a1a63b905576de41d26f/source/interaction/interaction_utilities.cc#L202-L269

**Comment by drwells** [comment]
# overview

@boyceg and I are going to use this issue to develop a roadmap for this project. If we want to integrate this with IBAMR then the best path forward is to use the same mechanisms we use for `fdl::IFEDMethod` to associate FE information (DoFHandlers, vectors, etc) with SAMRAI patches. All that association code is already written and reduces this project's scope considerably: once you have an `fdl::PatchMap` then you can use plain serial data structures. This makes development and testing pretty easy.

# roadmap

1. find all intersections between each finite difference stencil and the mesh
     a. in IBAMR, this is done element-by-element, with some bookkeeping to look only over a small subset of the mesh elements
     b. note that it is critical that intersections not be double-counted. in an element-by-element approach, we need to make sure we don't double-count intersections that occur on element edges or nodes
2. evaluate jump conditions at those intersections and evaluate the related correction terms, which appear as RHS terms in structured Cartesian grid equations 
    a. lowest order jump conditions are just the normal and tangential components of surface forces
    b. IBAMR currently uses the element geometry to describe the surface normals, but it might be better to use a globally continuous description of normal vectors
    c. it is not clear to me that we need to project the jump conditions onto basis functions that live on the finite element surface mesh; this is what was in Amin's initial implementation, but the DG scheme that Amin and Michael are using now is closer to just evaluating the jump conditions pointwise from the surface force F and the surface normal/tangential directions
3. tests
    a. we can do basic nontrivial tests just by applying forces on a stationary boundary and solving the N-S equations
    b. also we want to have unit tests for the basic intersection algorithms
    c. we can always start an IIM implementation using IB for velocity interpolation
    d. but there is no reason to do a fancy corrected velocity interpolation scheme until jump conditions are imposed in the N-S solver

# implementation

Most of the book-keeping work is already done by fiddle: see https://github.com/drwells/fiddle/blob/master/include/fiddle/interaction/ifed_method_base.h. That class can set things up and determine which elements intersect which patches.

During time-stepping, we need to take that information and compute the intersections.

**Comment by ryanleaf1000** [comment]
Hello David,

I am trying to build the band example and I keep getting this error:

/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::gzip_footer::reset()'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::do_init(boost::iostreams::zlib_params const&, bool, void* (*)(void*, unsigned int, unsigned int), void (*)(void*, void*), void*)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::zlib_base()'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `void dealii::DoFRenumbering::compute_support_point_wise<1, 2>(std::vector<unsigned long, std::allocator<unsigned long> >&, dealii::DoFHandler<1, 2> const&)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::best_compression'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::xinflate(int)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::gzip_header::process(char)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib_error::check(int)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::~zlib_base()'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::okay'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::before(char const*&, char const*, char*&, char*)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::best_speed'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::default_compression'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::default_strategy'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::reset(bool, bool)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::finish'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::gzip_header::reset()'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::no_flush'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::gzip_footer::process(char)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `void dealii::DoFRenumbering::compute_support_point_wise<2, 2>(std::vector<unsigned long, std::allocator<unsigned long> >&, dealii::DoFHandler<2, 2> const&)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::xdeflate(int)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::stream_end'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::detail::zlib_base::after(char const*&, char*&, bool)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::sync_flush'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::deflated'
collect2: error: ld returned 1 exit status
make[3]: *** [examples/elastic-band/CMakeFiles/band.dir/build.make:122: examples/elastic-band/main2d] Error 1
make[2]: *** [CMakeFiles/Makefile2:2240: examples/elastic-band/CMakeFiles/band.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2247: examples/elastic-band/CMakeFiles/band.dir/rule] Error 2
make: *** [Makefile:1144: band] Error 2

Do you know why?

**Comment by boyceg** [comment]
It looks like you may have an incomplete version of Boost. What boost are you using?On Jul 22, 2023, at 9:49 AM, Qi Sun ***@***.***> wrote:﻿
Hello David,
I am trying to build the band example and I keep getting this error:
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::gzip_footer::reset()' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::do_init(boost::iostreams::zlib_params const&, bool, void* ()(void, unsigned int, unsigned int), void ()(void, void*), void*)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::zlib_base()' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to void dealii::DoFRenumbering::compute_support_point_wise<1, 2>(std::vector<unsigned long, std::allocator >&, dealii::DoFHandler<1, 2> const&)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::best_compression' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::xinflate(int)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::gzip_header::process(char)' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib_error::check(int)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::~zlib_base()' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::okay'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::before(char const*&, char const*, char*&, char*)' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::best_speed'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::default_compression' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::default_strategy'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::reset(bool, bool)' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::finish'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::gzip_header::reset()' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::no_flush'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::gzip_footer::process(char)' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to void dealii::DoFRenumbering::compute_support_point_wise<2, 2>(std::vector<unsigned long, std::allocator >&, dealii::DoFHandler<2, 2> const&)'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::xdeflate(int)' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::stream_end'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::detail::zlib_base::after(char const*&, char*&, bool)' /usr/bin/ld: ../../libfiddle2d.so: undefined reference to boost::iostreams::zlib::sync_flush'
/usr/bin/ld: ../../libfiddle2d.so: undefined reference to `boost::iostreams::zlib::deflated'
collect2: error: ld returned 1 exit status
make[3]: *** [examples/elastic-band/CMakeFiles/band.dir/build.make:122: examples/elastic-band/main2d] Error 1
make[2]: *** [CMakeFiles/Makefile2:2240: examples/elastic-band/CMakeFiles/band.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2247: examples/elastic-band/CMakeFiles/band.dir/rule] Error 2
make: *** [Makefile:1144: band] Error 2
Do you know why?

**Comment by ryanleaf1000** [comment]
I am using boost 1.82. And IBAMR works fine.

**Comment by boyceg** [comment]
IBAMR uses only a very small part of Boost.

How did you install Boost for fiddle?

**Comment by ryanleaf1000** [comment]
I followed this instruction for building fiddle:
https://github.com/drwells/fiddle
I didn't do anything specific for boost.

**Comment by boyceg** [comment]
OK, but which Boost is being used?

**Comment by ryanleaf1000** [comment]
I followed the IBAMR instruction for BOOST:
https://ibamr.github.io/linux#boost
This is what I did except for the boost version 1.82.0
cd $HOME/sfw/linux
mkdir boost
cd boost
tar xvfz path/to/boost_1_60_0.tar.gz
mv boost_1_60_0 1.60.0
export BOOST_ROOT=$HOME/sfw/linux/boost/1.60.0
mkdir $BOOST_ROOT/include
ln -s $BOOST_ROOT/boost $BOOST_ROOT/include

**Comment by boyceg** [comment]
I’m not sure whether those instructions are sufficient for fiddle.

Does deal.II also use boost and are you using the same boost for everything?

It looks like you might actually need to build boost and not just unpack it — but I’m not sure.

**Comment by ryanleaf1000** [comment]
Thanks, Boyce! I installed boost. And this fixed the problem!

**Comment by drwells** [comment]
Glad you figured it out - I'll add another CMake check to try and detect this.

Could you open a new issue for this next time instead of posting here? This issue is for tracking IIM. Unless either of you object I'll remove the non-IIM posts at the end of the day.

**Comment by ryanleaf1000** [comment]
Hello David,

I pretty much finished testing intersection code. Here are the two functions that I want to add to fiddle:
  bool
  intersect_line_with_edge(std::vector<std::pair<double, Point<1>> >& t_vals,
                           const DoFHandler<1, 2>::active_cell_iterator elem,
                           const dealii::MappingQ<1, 2> &mapping,
                           dealii::Point<2> r,
                           dealii::Tensor<1,2> q,
                           const double tol = 0.0);
  bool
  intersect_line_with_face(std::vector<std::pair<double, Point<2>> >& t_vals,
                           const typename DoFHandler<2, 3>::active_cell_iterator elem,
                           const dealii::MappingQ<2, 3> &mapping,
                           dealii::Point<3> r,
                           dealii::Tensor<1,3> q,
                           const double tol=0);
Where would you prefer me to add these two functions to? grid_utilities.cc? In the two functions, I do loop over interface elements.

**Comment by drwells** [comment]
Thanks. How about we put these in `interaction/interaction_utilities.h` and open a pull request?

**Comment by ryanleaf1000** [comment]
Hello David,
I opened a pull request yesterday, although I don't think it's ready to be merged. I believe that there are likely still modifications needed for it to be completely ready for force spreading. But please feel free to share your comments, thoughts, or any preferences you would prefer me to modify?

**Comment by drwells** [comment]
Lets move that discussion over there - this issue is just the roadmap.

**Comment by drwells** [comment]
@ryanleaf1000 Looking at the roadmap, one next step is figuring out how to handle normal vectors. Another is populating `PatchIntersectionMap`. I'll add some notes on what we can do next for either one up above. Which one would you like to try?

**Comment by ryanleaf1000** [comment]
Either one could work. I am not sure I fully understand exactly what I  need to do to populating PatchIntersectionMap. Which one you think we should work on first?
Regarding handle normal vectors, I have some thought and some most recent findings to share with you.
1) we are still not sure if Phong's shading would solve the existing leaking problem. Maybe we should wait till Michael's tests with IBAMR and then decide? 
2) Recently I tried nodal split the forces for jump conditions. So far everything seems working fine at least for the existing examples in IBAMR. Do you think it's necessary for us to have a global representation of the normal instead we could have a global representation of the force and split it into jump condition only on intersection points and interpoation points on the interface? What do you think of directly using point wise normal? Or potentially using level set for calculating normal? What do you think?
3) I heard Amin is working on near touching problem and I also have some interpolation algorithm for two interfaces within one stencil. I know it's too early for this. I brought up this because maybe we could  design some potential extendable feature for storing intersection points from two interfaces? For sure this feature would be useful for double counted interpolation in future if we decide to do near touching problems which is one of advantage of IIM.

**Comment by drwells** [comment]
> I am not sure I fully understand exactly what I need to do to populating PatchIntersectionMap. Which one you think we should work on first?

One place to start with `PatchIntersectionMap` would be to sketch out the algorithms we need. We should probably use an `rtree` to look up elements which may intersect a given stencil. We can draft this on paper.

The normal vectors are probably an easier place to start. We just need a representation of normal vectors given some FE space; we can start by asserting that we have multicomponent piecewise constants.

> we are still not sure if Phong's shading would solve the existing leaking problem. Maybe we should wait till Michael's tests with IBAMR and then decide?

>  Do you think it's necessary for us to have a global representation of the normal?

I think we should represent the normal vectors as an FE field. That we we can do whatever we want with it later on.

> instead we could have a global representation of the force and split it into jump condition only on intersection points and interpoation points on the interface for now? Or potentially using level set for calculating normal? What do you think?

Doing that still requires having normal vectors available on each cell, so we need a way to compute those.

> I heard Amin is working on near touching problem and I also have some interpolation algorithm for two interfaces within one stencil. I know it's too early for this. I brought up this because maybe we could design some potential extendable feature for storing intersection points from two interfaces?

Doing this in parallel is going to be quite tricky so lets hold off on it for now and stick with single intersections.
